File: StockAnalyzer-real-stack/lib/models/stock.dart
class Stock {
  final int id;
  final String ticker;
  final String exchange;
  final String name;
  final String? country;
  final String? regionCode;
  final String? sector;
  final String? industry;
  final String? description;

  Stock({
    required this.id,
    required this.ticker,
    required this.exchange,
    required this.name,
    this.country,
    this.regionCode,
    this.sector,
    this.industry,
    this.description,
  });

  factory Stock.fromJson(Map<String, dynamic> json) {
    return Stock(
      id: json['id'] as int,
      ticker: json['ticker'] as String,
      exchange: json['exchange'] as String,
      name: json['name'] as String,
      country: json['country'] as String?,
      regionCode: json['region_code'] as String?,
      sector: json['sector'] as String?,
      industry: json['industry'] as String?,
      description: json['description'] as String?,
    );
  }
}

class ShariahStatus {
  final String status;
  final String grade;
  final double? debtRatio;
  final double? haramIncomeRatio;
  final double? purificationRate;
  final String explanation;
  final String? asOfDate;

  ShariahStatus({
    required this.status,
    required this.grade,
    this.debtRatio,
    this.haramIncomeRatio,
    this.purificationRate,
    required this.explanation,
    this.asOfDate,
  });

  factory ShariahStatus.fromJson(Map<String, dynamic> json) {
    return ShariahStatus(
      status: json['status'] as String,
      grade: json['grade'] as String,
      debtRatio: json['debt_ratio'] != null
          ? (json['debt_ratio'] as num).toDouble()
          : null,
      haramIncomeRatio: json['haram_income_ratio'] != null
          ? (json['haram_income_ratio'] as num).toDouble()
          : null,
      purificationRate: json['purification_rate'] != null
          ? (json['purification_rate'] as num).toDouble()
          : null,
      explanation: json['explanation'] as String,
      asOfDate: json['as_of_date'] as String?,
    );
  }

  // Color mapping for grades
  static int getGradeColor(String grade) {
    switch (grade) {
      case 'A':
        return 0xFF4CAF50; // Green
      case 'B':
        return 0xFFFFEB3B; // Yellow
      case 'C':
        return 0xFFFF9800; // Orange
      case 'D':
        return 0xFFFF5722; // Deep Orange/Red-Orange for Warning
      case 'F':
        return 0xFFF44336; // Red
      default:
        return 0xFF9E9E9E; // Grey
    }
  }
}

class Fundamentals {
  final double? totalAssets;
  final double? totalDebt;
  final double? cashAndEquiv;
  final double? totalRevenue;
  final double? interestIncome;
  final double? interestExpense;
  final double? netIncome;
  final double? dividendsPerShare;
  final String? asOfDate;
  final String? source;

  Fundamentals({
    this.totalAssets,
    this.totalDebt,
    this.cashAndEquiv,
    this.totalRevenue,
    this.interestIncome,
    this.interestExpense,
    this.netIncome,
    this.dividendsPerShare,
    this.asOfDate,
    this.source,
  });

  factory Fundamentals.fromJson(Map<String, dynamic> json) {
    return Fundamentals(
      totalAssets: json['total_assets'] != null
          ? (json['total_assets'] as num).toDouble()
          : null,
      totalDebt: json['total_debt'] != null
          ? (json['total_debt'] as num).toDouble()
          : null,
      cashAndEquiv: json['cash_and_equiv'] != null
          ? (json['cash_and_equiv'] as num).toDouble()
          : null,
      totalRevenue: json['total_revenue'] != null
          ? (json['total_revenue'] as num).toDouble()
          : null,
      interestIncome: json['interest_income'] != null
          ? (json['interest_income'] as num).toDouble()
          : null,
      interestExpense: json['interest_expense'] != null
          ? (json['interest_expense'] as num).toDouble()
          : null,
      netIncome: json['net_income'] != null
          ? (json['net_income'] as num).toDouble()
          : null,
      dividendsPerShare: json['dividends_per_share'] != null
          ? (json['dividends_per_share'] as num).toDouble()
          : null,
      asOfDate: json['as_of_date'] as String?,
      source: json['source'] as String?,
    );
  }

  // Format large numbers (billions, millions)
  static String formatCurrency(double? value) {
    if (value == null) return 'N/A';
    if (value >= 1000000000) {
      return '\$${(value / 1000000000).toStringAsFixed(2)}B';
    } else if (value >= 1000000) {
      return '\$${(value / 1000000).toStringAsFixed(2)}M';
    } else if (value >= 1000) {
      return '\$${(value / 1000).toStringAsFixed(2)}K';
    }
    return '\$${value.toStringAsFixed(2)}';
  }
}

File: backend/internal/models/stock.go
package models

import (
	"database/sql"
	"time"
)

type Stock struct {
	ID          int64
	Ticker      string
	Exchange    string
	Name        string
	Country     sql.NullString
	RegionCode  sql.NullString
	Sector      sql.NullString
	Industry    sql.NullString
	Description sql.NullString
	MarketCap   sql.NullInt64
	IsActive    bool
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

type Fundamental struct {
	ID                int64
	StockID           int64
	TotalAssets       sql.NullFloat64
	TotalDebt         sql.NullFloat64
	CashAndEquiv      sql.NullFloat64
	TotalRevenue      sql.NullFloat64
	InterestIncome    sql.NullFloat64
	InterestExpense   sql.NullFloat64
	NetIncome         sql.NullFloat64
	DividendsPerShare sql.NullFloat64
	AsOfDate          sql.NullTime
	Source            sql.NullString
	RawJSON           sql.NullString
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

type ShariahStatus struct {
	ID               int64
	StockID          int64
	Status           string         // HALAL, HARAM, MIXED, UNKNOWN
	Grade            sql.NullString // A, B, C, F
	DebtRatio        sql.NullFloat64
	HaramIncomeRatio sql.NullFloat64
	PurificationRate sql.NullFloat64
	PaysZakat        sql.NullBool
	Explanation      sql.NullString
	Reason           sql.NullString
	AsOfDate         time.Time
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

type User struct {
	ID                  int64
	Email               string
	PasswordHash        string
	Name                sql.NullString
	// AvatarURL — profile image. Populated by PATCH /me/profile (mig 0038).
	// Stays NULL until the user uploads an avatar; clients fall back to
	// a gradient + initials when empty.
	AvatarURL           sql.NullString
	Role                string         // USER, EXPERT, SCHOLAR (default USER)
	ExpertID            sql.NullString // FK → experts.id when role != USER
	SubscriptionTier    string         // FREE, PREMIUM
	SubscriptionStatus  string         // ACTIVE, CANCELED, EXPIRED
	SubscriptionEndDate sql.NullTime
	// Locale — UI language synced from the app ('en' / 'ar'). Drives
	// per-recipient notification push localization. NOT NULL default 'en'.
	Locale              string
	CreatedAt           time.Time
	UpdatedAt           time.Time
}

type UserPortfolio struct {
	ID          int64
	UserID      int64
	StockID     int64
	Shares      sql.NullFloat64
	AvgBuyPrice sql.NullFloat64
	AddedAt     time.Time
}

type Notification struct {
	ID        int64
	UserID    sql.NullInt64
	StockID   sql.NullInt64
	Type      string
	Title     string
	Message   string
	IsRead    bool
	CreatedAt time.Time
}

type AnalystRating struct {
	ID          int64
	StockID     int64
	AnalystName sql.NullString
	Rating      sql.NullString // BUY, HOLD, SELL
	TargetPrice sql.NullFloat64
	RatingDate  sql.NullTime
	Source      sql.NullString
	CreatedAt   time.Time
}

// Ad — one row in the `ads` table. Used both by the public
// /api/ads?region_code=… endpoint and by the admin CRUD. JSON tags use
// the camelCase shape the admin dashboard + Flutter clients expect.
type Ad struct {
	ID          int64          `json:"id"`
	CompanyName string         `json:"companyName"`
	Title       string         `json:"title"`
	Description sql.NullString `json:"-"`
	ImageURL    sql.NullString `json:"-"`
	TargetURL   sql.NullString `json:"-"`
	RegionCode  sql.NullString `json:"-"`
	IsActive    bool           `json:"isActive"`
	StartDate   sql.NullTime   `json:"-"`
	EndDate     sql.NullTime   `json:"-"`
	CreatedAt   time.Time      `json:"createdAt"`
}

// AdJSON is the wire shape — collapses the sql.Null* wrappers into
// nullable pointer fields so the client sees `null` instead of
// `{String:"", Valid:false}`. The repo returns these for HTTP paths.
type AdJSON struct {
	ID          int64      `json:"id"`
	CompanyName string     `json:"companyName"`
	Title       string     `json:"title"`
	Description *string    `json:"description,omitempty"`
	ImageURL    *string    `json:"imageUrl,omitempty"`
	TargetURL   *string    `json:"targetUrl,omitempty"`
	RegionCode  *string    `json:"regionCode,omitempty"`
	IsActive    bool       `json:"isActive"`
	StartDate   *time.Time `json:"startDate,omitempty"`
	EndDate     *time.Time `json:"endDate,omitempty"`
	CreatedAt   time.Time  `json:"createdAt"`
}

// ToJSON unwraps the sql.Null* fields. Empty strings/zero times remain
// non-nil if the column was set to "" rather than NULL — that's
// intentional, the admin's "clear" flow uses NULL.
func (a *Ad) ToJSON() AdJSON {
	out := AdJSON{
		ID:          a.ID,
		CompanyName: a.CompanyName,
		Title:       a.Title,
		IsActive:    a.IsActive,
		CreatedAt:   a.CreatedAt,
	}
	if a.Description.Valid {
		s := a.Description.String
		out.Description = &s
	}
	if a.ImageURL.Valid {
		s := a.ImageURL.String
		out.ImageURL = &s
	}
	if a.TargetURL.Valid {
		s := a.TargetURL.String
		out.TargetURL = &s
	}
	if a.RegionCode.Valid {
		s := a.RegionCode.String
		out.RegionCode = &s
	}
	if a.StartDate.Valid {
		t := a.StartDate.Time
		out.StartDate = &t
	}
	if a.EndDate.Valid {
		t := a.EndDate.Time
		out.EndDate = &t
	}
	return out
}

// PromoCode — one row in promo_codes. Same JSON-tag pattern as Ad.
type PromoCode struct {
	ID            int64         `json:"id"`
	Code          string        `json:"code"`
	DiscountType  string        `json:"discountType"` // PERCENTAGE, FIXED
	DiscountValue float64       `json:"discountValue"`
	// Scope — which kind of purchase the code applies to:
	// "all" (default), "expert", or "community".
	Scope         string        `json:"scope"`
	MaxUses       sql.NullInt64 `json:"-"`
	UsedCount     int64         `json:"usedCount"`
	IsActive      bool          `json:"isActive"`
	ValidFrom     sql.NullTime  `json:"-"`
	ValidUntil    sql.NullTime  `json:"-"`
	CreatedAt     time.Time     `json:"createdAt"`
}

// PromoCodeJSON — wire shape for admin endpoints.
type PromoCodeJSON struct {
	ID            int64      `json:"id"`
	Code          string     `json:"code"`
	DiscountType  string     `json:"discountType"`
	DiscountValue float64    `json:"discountValue"`
	Scope         string     `json:"scope"`
	MaxUses       *int64     `json:"maxUses,omitempty"`
	UsedCount     int64      `json:"usedCount"`
	IsActive      bool       `json:"isActive"`
	ValidFrom     *time.Time `json:"validFrom,omitempty"`
	ValidUntil    *time.Time `json:"validUntil,omitempty"`
	CreatedAt     time.Time  `json:"createdAt"`
}

func (p *PromoCode) ToJSON() PromoCodeJSON {
	scope := p.Scope
	if scope == "" {
		scope = "all"
	}
	out := PromoCodeJSON{
		ID:            p.ID,
		Code:          p.Code,
		DiscountType:  p.DiscountType,
		DiscountValue: p.DiscountValue,
		Scope:         scope,
		UsedCount:     p.UsedCount,
		IsActive:      p.IsActive,
		CreatedAt:     p.CreatedAt,
	}
	if p.MaxUses.Valid {
		v := p.MaxUses.Int64
		out.MaxUses = &v
	}
	if p.ValidFrom.Valid {
		t := p.ValidFrom.Time
		out.ValidFrom = &t
	}
	if p.ValidUntil.Valid {
		t := p.ValidUntil.Time
		out.ValidUntil = &t
	}
	return out
}
